src/hotspot/share/opto/castnode.cpp

```diff
@@ -246,6 +246,38 @@ Node *CastIINode::Ideal(PhaseGVN *phase, bool can_reshape) {
     return progress;
   }
 
+  PhaseIterGVN *igvn = phase->is_IterGVN();
+  const TypeInt* this_type = this->type()->is_int();
+  Node* z = in(1);
+  const TypeInteger* rx = NULL;
+  const TypeInteger* ry = NULL;
+  // Similar to ConvI2LNode::Ideal() for the same reasons
+  if (!_range_check_dependency && Compile::push_thru_add(phase, z, this_type, rx, ry, T_INT)) {
+    if (igvn == NULL) {
+      // Postpone this optimization to iterative GVN, where we can handle deep
+      // AddI chains without an exponential number of recursive Ideal() calls.
+      phase->record_for_igvn(this);
+      return NULL;
+    }
+    int op = z->Opcode();
+    Node* x = z->in(1);
+    Node* y = z->in(2);
+
+    Node* cx = new CastIINode(x, rx->is_int(),
+                              _carry_dependency, _range_check_dependency);
+    cx->set_req(0, in(0));
+    cx = phase->transform(cx);
```

Comment:
Hi Roland, have you tested that these calls to `transform()` do not create an explosion of recursive `Ideal()` calls in the face of long AddI chains (see https://github.com/openjdk/jdk/pull/727).

Reply:
I wondered about that but had no issue during testing. Let me see if I can change your test case to have recursive Ideal() calls.

Reply:
I managed to reproduce a similar issue by tweaking your test cases. I just pushed the new test cases + a change that mirrors your fix.

Reply:
Great, thanks!